src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -449,8 +450,12 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // ptr and offset_ptr must be in bounds of the same allocated object. This means all of the
         // memory between these pointers must be accessible. Note that we do not require the
         // pointers to be properly aligned (unlike a read/write operation).
-        let min_ptr = if offset_bytes >= 0 { ptr } else { offset_ptr };
-        let size = offset_bytes.checked_abs().unwrap();
+        let (min_ptr, size) = if offset_bytes >= 0 {
+            (ptr, offset_bytes as u64)
+        } else {
+            // Compute (-offset_bytes as u64), but handle (offset_bytes == isize::MAX) case.
+            (offset_ptr, 0u64.wrapping_sub(offset_bytes as u64))
```

Comment:
Same here, please avoid `as`.

Reply:
In fact `offset_bytes` is negative here so what is even going on?

Reply:
This was based on how we handle computing `abs(n: isize) -> usize` with `n` negative elsewhere in `librustc_middle`: https://github.com/rust-lang/rust/blob/9eedd138ee22147111a885d6948fb050d9849bf4/src/librustc_middle/mir/interpret/pointer.rs#L52-L54

Using `as` to convert `i64` to `u64` wraps (i.e. takes the value mod 2^64) so doing a wrapping subtract after that operation computes the right value. 

It's possible to do this without an `as`, but all the alternatives are way less readable.

Reply:
This could also be written as `offset_bytes.wrapping_abs() as u64`, which look simpler but hides the subtleties. Namely that `i64::MAX.wrapping_abs() == i64::MAX`, but then `i64::MAX as u64` then wraps to `1 << 63`, the correct value.

Reply:
Yeah I think I understood it now... could you please factor out that non-wrapping `abs` operation of type `fn(i64) -> u64` into a separate function, with suitable comments? I think that would make the code here much easier to follow.

Reply:
> Namely that i64::MAX.wrapping_abs() == i64::MAX, but then i64::MAX as u64 then wraps to 1 << 63, the correct value.

I think you mean `MIN` here?

Reply:
> This was based on how we handle computing abs(n: isize) -> usize with n negative elsewhere in librustc_middle: 

If we also have such code there, even more of a reason to make that `abs` function a helper function that we can share. This should probably be in libstd, but until then... maybe add it as `uabs` for the bottom of `librustc_middle/mir/interpret/mod.rs` or so?

Reply:
I added `uabs` to `librustc_middle/mir/interpret/mod.rs`. I wasn't able to find any other obvious call sites, though I'm sure some more exist somewhere in the tree.